services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/Wrappers.groovy

```diff
@@ -0,0 +1,93 @@
+
+import groovy.util.logging.Slf4j
+import org.junit.runners.model.MultipleFailureException
+
+@Slf4j
+class Wrappers {
+
+    /**
+     * Retries some operation for a given number of times. Retry is considered failed if any throwable is thrown.
+     * Throws MultipleFailureException if closure fails to execute in given amount of tries.
+     *
+     * @param times number of retries
+     * @param retryInterval pause between retries (in seconds)
+     * @param handler exception handler (receives exception on each failure)
+     * @param body operation to wrap around
+     * @return result of the body execution
+     */
+    def static retry(int times = 5, double retryInterval = 2, Closure handler = { e -> log.debug(e) }, Closure body) {
+        int retries = 0
+
+        List<Throwable> ex = []
+        while (retries++ < times) {
+            try {
+                return body.call()
+            } catch (Throwable t) {
+                ex.add(t)
+                handler.call(t)
+                sleep(retryInterval * 1000 as long)
+            }
+        }
+        throw new MultipleFailureException(ex)
+    }
+
+    /**
+     * Wait for closure to return a result. Returns false if success condition was not met.
+     *
+     * @param timeout time to wait in seconds
+     * @param retryInterval pause between retries (in seconds)
+     * @param closure : A closure that returns following:
+     *          - True if check is successful
+     *          - False or null if check is unsuccessful
+     * @return True if wait was successful, false otherwise
+     */
+    static boolean wait(double timeout, double retryInterval = 0.5, Closure closure) {
+        long endTime = System.currentTimeMillis() + (long) (timeout * 1000)
+        long sleepTime = (long) (retryInterval * 1000)
+        while (System.currentTimeMillis() < endTime) {
+            try {
+                if (closure.call()) {
+                    return true
+                } else {
+                    log.debug("No wait results yet. Sleeping for ${sleepTime} ms...")
+                    sleep(sleepTime)
+                }
+            } catch (Throwable t) {
+                log.debug("Wait check threw exception: ${t.message}.\nSleeping for ${sleepTime} ms...")
+                sleep(sleepTime)
+            }
+        }
+        log.debug("Closure failed to succeed within ${timeout} seconds, aborting.")
+        return false
+    }
+
+    /**
+     * Debug only. Logs how much time certain code block took to execute
+     *
+     * @param name some name of closure/code block that is being executed
+     * @param closure code block to benchmark
+     * @return result of the given closure
+     */
+    static def benchmark(name, closure) {
+        def start = System.currentTimeMillis()
+        def result = closure.call()
+        def now = System.currentTimeMillis()
+        log.debug("$name took " + (now - start) + "ms")
+        result
```

Comment:
Sometimes uses return, sometimes not. Was it done specially?

Reply:
In Groovy 'return' keyword is optional, but I agree that for such large blocks it is better to explicitly mention it. Will add

Reply:
Yes, I know. Just to ask)